File: problems/0005_longest-palindromic-substring/central-expansion/solution.go
func longestPalindrome(s string) string {
    sLen := len(s)
    iEnd := sLen * 2 - 1
    maxLeft, maxRight := 0, 0
    for i := 0; i < iEnd; i += 1 {
        posLeft := i / 2
        posRight := posLeft
        if i % 2 == 1 {
            posRight += 1
        }
        for 0 <= posLeft && posRight < sLen &&
                s[posLeft] == s[posRight] {
            posLeft -= 1
            posRight += 1
        }
        posLeft += 1
        if posRight - posLeft > maxRight - maxLeft {
            maxLeft, maxRight = posLeft, posRight
        }
    }
    return s[maxLeft:maxRight]
}

File: problems/0005_longest-palindromic-substring/central-expansion/Solution.java
class Solution {
    public String longestPalindrome(String s) {
        final int l = s.length(),
                  iEnd = l * 2 - 1;
        int maxLeft = 0, maxRight = 0;
        for (int i = 0; i < iEnd; ++i) {
            int posLeft = i / 2,
                posRight = (i % 2 == 1 ? posLeft + 1 : posLeft);
            while (0 <= posLeft && posRight < l &&
                    s.charAt(posLeft) == s.charAt(posRight)) {
                --posLeft;
                ++posRight;
            }
            if (posRight - (++posLeft) > maxRight - maxLeft) {
                maxLeft = posLeft;
                maxRight = posRight;
            }
        }
        return s.substring(maxLeft, maxRight);
    }
}
